internal/witness/witness.go

```diff
@@ -196,16 +207,30 @@ func (w *witness) update(ctx context.Context, cp []byte, size uint64) ([]byte, e
 	body += "\n"
 	body += string(cp)
 
-	resp, err := w.post(ctx, w.url, body)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, w.url, strings.NewReader(body))
+	if err != nil {
+		return nil, fmt.Errorf("failed to post to witness at %q: %v", w.url, err)
+	}
+	httpResp, err := w.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to post to witness at %q: %v", w.url, err)
+	}
+	rb, err := io.ReadAll(httpResp.Body)
 	if err != nil {
 		return nil, fmt.Errorf("failed to post to witness at %q: %v", w.url, err)
 	}
+	_ = httpResp.Body.Close()
 
-	switch resp.statusCode {
+	switch httpResp.StatusCode {
 	case http.StatusOK:
-		signed := append(cp, resp.body...)
+		// Concatenate the signature to the checkpoint passed in and verify it is valid.
+		// append is tempting here but is dangerous because it can modify `cp` and race with other
+		// witnesses, causing signatures to be swapped. cp must not be modified.
+		signed := make([]byte, len(cp)+len(rb))
+		copy(signed, cp)
+		copy(signed[len(cp):], rb)
 		if n, err := note.Open(signed, note.VerifierList(w.verifier)); err != nil {
-			return nil, fmt.Errorf("witness at %q replied with invalid signature: %q", w.url, resp.body)
+			return nil, fmt.Errorf("witness %q at %q replied with invalid signature: %q\nconstructed note: %q\n%v", w.verifier.Name(), w.url, rb, string(signed), err)
```

Comment:
nit: might be easy to miss the error message in the log - worth modifying to `...note: %q\nerror: %v`?

Reply:
Done.